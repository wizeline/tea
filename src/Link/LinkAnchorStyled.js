// @flow

import styled from 'styled-components';

const LinkAnchorStyled = styled.a`
  text-decoration: none !important;
`;

export default LinkAnchorStyled;
