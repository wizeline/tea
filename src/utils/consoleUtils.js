// @flow

const warn = (warning: string) => {
  /* eslint-disable no-console */
  console.warn(warning);
  /* eslint-enable no-console */
  return null;
};

export default warn;
