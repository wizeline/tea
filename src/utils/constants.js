// @flow

const multipleTruthyKeyError =
  'ERROR: Multiple keys associated with a truthy value, offending keys:';

const supportedPropNames = ['size', 'name', 'color', 'background'];
const supportedCssProps = ['background', 'fill', 'height', 'top', 'width'];

export { multipleTruthyKeyError, supportedCssProps, supportedPropNames };
