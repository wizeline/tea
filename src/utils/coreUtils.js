// @flow

import reduce from 'ramda/src/reduce';
import { css } from 'styled-components';
import {
  multipleTruthyKeyError,
  supportedCssProps,
  supportedPropNames,
} from './coreUtilsConstants';

const camelToLispCase = (word: string) =>
  word.replace(/([A-Z])/g, captureGroup => `-${captureGroup.toLowerCase()}`);

const getTruthyKey = (obj: Object): ?string => {
  const truthyKeys = Object.keys(obj).filter(key => Boolean(obj[key]));
  if (truthyKeys.length > 1) {
    const reason = truthyKeys.map(key => `'${key}'`).join(', ');
    const errorMessage = `${multipleTruthyKeyError} [${reason}]`;
    throw new Error(errorMessage);
  }
  return truthyKeys[0];
};

const curryPropParsers = (...parsers: Array<Function>) => (props: Object) => {
  const reducer = (accumulator, currentValue) => ({
    ...accumulator,
    ...currentValue,
  });
  return parsers.map(parser => parser(props)).reduce(reducer, {});
};

const getObjectSupport = (names: Array<string>, mapping?: Function): Object => {
  const result = {};
  names.forEach((name: string) => {
    result[name] = mapping ? mapping(name) : name;
  });
  return result;
};

const cssProperties = getObjectSupport(supportedCssProps, camelToLispCase);
const propNames = getObjectSupport(supportedPropNames);

const propOrTheme = (themeName: string, propName: string): Function => (
  props: Object,
) => (props[propName] ? props[propName] : props.theme[themeName]);

const toPixels = (val: any) => (val ? `${val}px` : '0px');

const supportOr = (
  input: string,
  defaultOutput: any,
  mapping: Function,
  ...outputs: Array<string>
) => (props: Object) => {
  const value = props[input] || defaultOutput;
  const mappedValue = mapping(value);
  const targets = outputs.length ? outputs : [input];
  const generated = targets
    .map(target => `${target}: ${mappedValue};`)
    .join('\n');

  return css`
    ${generated};
  `;
};

const supportOrTheme = (input: string, defaultThemePropName: string) => (
  props: Object,
) => {
  const propertyValue = props[input] || props.theme[defaultThemePropName];
  const generated = [input]
    .map(propertyName => `${propertyName}: ${propertyValue};`)
    .join('\n');

  return css`
    ${generated};
  `;
};

const warn = (warning: string) => {
  /* eslint-disable no-console */
  console.warn(warning);
  /* eslint-enable no-console */
};

const supportBooleanNameGroup = (groupName: string, names: Array<string>) => (
  props: Object,
) => {
  const activeName = names.find(name => props[name]);
  return activeName ? { [groupName]: activeName } : {};
};

const arrayToObject = (array: Array<string>) => {
  const accumulator = (a, b) => ({ ...a, [b]: true });
  return reduce(accumulator, {}, array);
};

export {
  arrayToObject,
  cssProperties,
  curryPropParsers,
  getTruthyKey,
  camelToLispCase,
  propNames,
  propOrTheme,
  supportBooleanNameGroup,
  supportOr,
  supportOrTheme,
  toPixels,
  warn,
};
