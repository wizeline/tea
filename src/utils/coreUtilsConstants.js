// @flow

const multipleTruthyKeyError =
  'ERROR: Multiple keys associated with a truthy value, offending keys:';

const supportedCssProps = ['top', 'background', 'fill'];

export { multipleTruthyKeyError, supportedCssProps };
