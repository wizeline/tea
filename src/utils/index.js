// @flow

import supportSizes from './sizes/supportSizes';
import {
  arrayToObject,
  cssProperties,
  lispToCamelCase,
  propNames,
  propOrTheme,
  supportBooleanNameGroup,
  supportOr,
  supportOrTheme,
  toPixels,
  warn,
} from './coreUtils';

export {
  arrayToObject,
  cssProperties,
  lispToCamelCase,
  propNames,
  propOrTheme,
  supportBooleanNameGroup,
  supportOr,
  supportOrTheme,
  supportSizes,
  toPixels,
  warn,
};
