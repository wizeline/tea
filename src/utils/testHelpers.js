// @flow

/* eslint flowtype-errors/enforce-min-coverage: 0 */

import * as React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import toJson from 'enzyme-to-json';
import ThemeProvider from '../ThemeProvider';

const { expect } = global;

const shallowExpectInvalid = (component: React.Node) => {
  const wrapper: ShallowWrapper = shallow(component);
  expect(toJson(wrapper)).toMatchSnapshot();
  expect(toJson(wrapper)).toBe('');
};

const shallowExpectToThrow = (component: React.Node) => {
  expect(() => {
    shallow(component);
  }).toThrow();
};

const shallowExpectToMatchSnapShot = (component: React.Node) => {
  const wrapper: ShallowWrapper = shallow(component);
  expect(toJson(wrapper)).toMatchSnapshot();
};

const shallowWithTheme = (children: ?React.Node, theme?: Object) => {
  const wrapper = shallow(
    <ThemeProvider theme={theme}>{children}</ThemeProvider>,
  );
  const instance = wrapper.dive().instance();
  return wrapper.shallow({ context: instance.getChildContext() });
};

export {
  shallowExpectInvalid,
  shallowExpectToMatchSnapShot,
  shallowExpectToThrow,
  shallowWithTheme,
};
