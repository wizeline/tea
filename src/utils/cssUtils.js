// @flow

import { css } from 'styled-components';
import { getObjectSupport } from './objectUtils';
import { supportedCssProps } from './constants';
import { camelToLispCase } from './stringUtils';

const toPixels = (val: any) => (val ? `${val}px` : '0px');
const cssProperties = getObjectSupport(supportedCssProps, camelToLispCase);

const supportOr = (
  input: string,
  defaultOutput: any,
  mapping: Function,
  ...outputs: Array<string>
) => (props: Object) => {
  const value = props[input] || defaultOutput;
  const mappedValue = mapping(value);
  const targets = outputs.length ? outputs : [input];
  const generated = targets
    .map(target => `${target}: ${mappedValue};`)
    .join('\n');

  return css`
    ${generated};
  `;
};

const supportOrTheme = (input: string, defaultThemePropName: string) => (
  props: Object,
) => {
  const propertyValue = props[input] || props.theme[defaultThemePropName];
  const generated = [input]
    .map(propertyName => `${propertyName}: ${propertyValue};`)
    .join('\n');

  return css`
    ${generated};
  `;
};

export { cssProperties, toPixels, supportOr, supportOrTheme };
