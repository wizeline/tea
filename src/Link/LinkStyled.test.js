import React from 'react';
import LinkAnchorStyled from './LinkStyled';
import { shallowExpectToMatchSnapShot } from '../utils/testHelpers';

describe('LinkAnchorStyled', () => {
  it('renders correctly with defined theme', () => {
    const childrenString = 'hello';
    const props = {
      href: 'http://wizeline.com/',
      priority: 'b100',
      children: childrenString,
      theme: {
        textLink: 'yellow',
      },
    };
    shallowExpectToMatchSnapShot(<LinkAnchorStyled {...props} />);
  });
  it('renders correctly with undefined theme', () => {
    const childrenString = 'hello';
    const props = {
      href: 'http://wizeline.com/',
      priority: 'b100',
      children: childrenString,
    };
    shallowExpectToMatchSnapShot(<LinkAnchorStyled {...props} />);
  });
});
