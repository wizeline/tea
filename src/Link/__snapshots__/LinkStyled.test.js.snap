// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinkAnchorStyled renders correctly with defined theme 1`] = `
.c0 {
  text-decoration: none;
  font-size: 14px;
  color: yellow;
}

<a
  className="c0"
  href="http://wizeline.com/"
>
  hello
</a>
`;

exports[`LinkAnchorStyled renders correctly with undefined theme 1`] = `
.c0 {
  text-decoration: none;
  font-size: 14px;
}

<a
  className="c0"
  href="http://wizeline.com/"
>
  hello
</a>
`;
